### ISSUE TITLE
Render ghost state differently?

### ISSUE THREAD

**Issue by robin-aws** [dafny-lang/ide-vscode#53]
I know this has come up before in discussion on these repos, but at the risk of creating a duplicate I wanted to mention it on this specific repo.

One common misunderstanding when first learning Dafny is that it's going to be super slow at runtime, because folks don't realize that ghost state is erased when compiling Dafny. There's also the fact that the standard library (https://github.com/dafny-lang/libraries) has an idiom of prefixing all lemmas with `Lemma...` to emphasize that these "methods" only have an effect on ghost state.

It could be awesome to have a mode where ghost state is rendered differently, perhaps greyed out. I could even imagine someone wanting to slice it all away entirely in order to only read the compiled aspect of a Dafny program.

**Comment by camrein** [comment]
You've got me way too excited with this feature request since I knew that all the necessary information is already present in the language server. I quickly prototyped an implementation for this that currently only fades out the declarations of ghosts:

![grafik](https://user-images.githubusercontent.com/18049310/135708644-2fbc0734-7e46-4abd-8b65-09b0ec3f1e32.png)

This feature can be solved solely with an extension to the language server. For each ghost, it has to publish a diagnostic with the tag `DiagnosticTag.Uncessary` set. https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#diagnosticTag

**Comment by robin-aws** [comment]
Wow Christoph, you rock! :) I'm super happy it's a lot easier to implement than I thought, but I agree we should put some thought into the UX here. Would it be easy to have hover text explaining why it's greyed out, and perhaps something like "press *** to disable this feature"?

**Comment by camrein** [comment]
I'm happy that you like it. :)
We get the hover text basically for free:
![image](https://user-images.githubusercontent.com/18049310/135745013-9219e84a-df0f-47fe-97c7-45868a7449d7.png)
![image](https://user-images.githubusercontent.com/18049310/135745033-cdbf04d9-290f-4529-804e-70455faf6884.png)
![image](https://user-images.githubusercontent.com/18049310/135745190-c3ad76fa-a6ec-4436-8ebd-92fdaf982e09.png)
However, I've had the feeling that hovering over designators inside the ghost function got a bit difficult. It appears to me that if there is already hover information displayed, it won't update what's displayed until it's hidden again. For example, if you hover near a variable and it starts showing "Ghost function" and move over the variable right after, it won't update the displayed text. So maybe we shouldn't grey out the whole function when we can't resolve this behavior with reasonable effort.

Enabling/disabling this feature on-request needs a bit more effort. So far, all language server options are passed upon startup as CLI arguments. That's why some of the extension options have the "requires restart" note. Supporting interactive switching will require that the language server exposes an API to update these options at runtime. If we do this, I guess it makes sense to support configuration updates without restarts for all options (where possible).

**Comment by robin-aws** [comment]
Glad we're experimenting here, I agree the greying-out approach is likely not the best option.

I also have to say that even though I don't think the last option is right either, those tiny little ghosts are ADORABLE. :)

**Comment by MikaelMayer** [comment]
Robin said
> One common misunderstanding when first learning Dafny is that it's going to be super slow at runtime, because folks don't realize that ghost state is erased when compiling Dafny.

Today, I had a great discussion with someone totally new to Dafny, walking them through features of the plugin. One of the interesting things for him was to understand that some portions of the code can be ghost (erased at compile-time) and compiled (kept). I had to do some extra work explaining him what was ghost and what was not, since there are so many keywords to indicate if something is ghost or not. So having a visual indicator of ghostness or not would definitely be a great thing.

There is a problem with the approach of only highlighting statements or sub-expressions that are ghost, and not things that knowledgeable Dafny users know to be always ghost.  
I envision the following conversation happening over and over:
    "- See, these statements are highlighted because they will be erased at compile time
    - Oh, so this highlight means it will be erased at compile time. I thought assertions would also be erased at compile time, but they are not highlighted, so they are actually compiled, right?
    - Err, hum, no, assertions are always ghost and since we all know that they are that's why we don't highlight them.
    - I wasn't sure of that. What are all the things that will be erased at compile time that are not highlighted?
    - Functions, predicates which are not marked as compiled, lemmas, pre-conditions, post-conditions, the main body of function by method but not their associated methods... let me think of others...
    - Ah, that's a lot to remember, I'll try."

In short, to give a consistent view and make the meaning of the highlighted code very clear to grasp, we should highlight all the portions that are going to be erased by most compilers.

**Comment by keyboardDrummer** [comment]
> In short, to give a consistent view and make the meaning of the highlighted code very clear to grasp, we should highlight all the portions that are going to be erased by most compilers.

Probably easiest way to decide would be to see several mock-ups.

I think it could makes sense to only apply the ghostness highlighting at the call site, not at the entire definition site for functions/methods that are ghost, although this relates to the discussion about ghostness of functions, what the default should be, or if they should ever be ghost.

**Comment by MikaelMayer** [comment]
It looks like the overwhelming initial consensus for now is that everyone wants to have some kind of ghost highlighting. Now the goal of this ghost highlighting can differ subtly, and we need to determine which goal suits which need.

I can easily imagine recent users appreciating to see everything ghost as ghost (to ease the learning curve, switching from other programming languages), and advanced users only wanting to see ghost markup where there is ambiguity, especially is the ghost markup was too prominent like in the screenshots above.

I'd love to have your opinion on subtle cases. Let us consider the case non-ghost variables in ghost expressions, like in the `ensures` clause above :

    ensures f1 == fibonacci(n)

The `ensures` is always ghost, the function `fibonacci` is ghost (but that was ambiguous), and n and f1 are not ghost (and that was ambiguous). The == sign in a ghost statement can fairly be viewed as ghost.
Therefore, I see many ways to highlight this with ghost markup, there "G" stands for ghost markup (whatever we prefer), "-" stands for a hybrid markup (something which exists in the compiler but is referenced to in a ghost context), and "_" stands for no highlight.

|Tag|Meaning|highlight|
|-----|----------|----------|
|Not compiled| Ghost highlight means not compiled | `ensures f1 == fibonacci(n)`<br>`GGGGGGGGGGGGGGGGGGGGGGGGGG`|
|Hybrid not compiled| Ghost+Hybrid highlight means not compiled, Hybrid highlight means something is not ghost usually | `ensures f1 == fibonacci(n)`<br>`GGGGGGGG--GGGGGGGGGGGGGG-G`|
|Ghost resolved| Ghost highlight means it's ghost by definition | `ensures f1 == fibonacci(n)`<br>`GGGGGGGG__GGGGGGGGGGGGGG_G`|
|Hybrid not compiled ghost resolved| Hybrid+Ghost highlight means not compiled, ghost highlight  means ambiguity has been resolved and it's ghost | `ensures f1 == fibonacci(n)`<br>`--------------GGGGGGGGGG-G`|
|Ambiguous Ghost resolved| Ghost means ambiguity has been resolved and it's ghost | `ensures f1 == fibonacci(n)`<br>`______________GGGGGGGGGG_G`|

What do you think of these cases, for different users (including yourself)?

**Comment by camrein** [comment]
Thanks for your sophisticated input.

> There is a problem with the approach of only highlighting statements or sub-expressions that are ghost, and not things that knowledgeable Dafny users know to be always ghost.

That supports my initial impression that this feature should be primarily purposed for beginner users. Therefore, I initially opted to mark everything as ghost. However, since I had the feeling that this might be too distracting, I made the levels configurable, i.e., include declarations.

Your sketches are great @MikaelMayer and beneficial since, compared to me, you'll definitely have more input from both beginner and advanced users.
But I have the feeling that highlighting hybrid differently to regular ghost might make the visual experience too verbose, for example, if we think back about Dafny's info messages (the blue squiggles). What about taking the route of only denoting ghost-state and letting users decide if they think it's helpful to see the hybrid state as well? This way, we avoid introducing code complexity within the language server with a feature that might be unnecessary. Furthermore, we could aim to introduce the ghost state feature with Dafny 3.4 since most of the things are have already been implemented.

**Comment by MikaelMayer** [comment]
Thanks for these thoughts. I'd say, let's aim at something simple first, and add complexity as needed later. For now, I think that first-time plugin users should see the highlighting above labeled "Not compiled" by default, and it should be possible to offer a "disambiguation highlighting" (the one everyone else introduces here), label ghost statements and expressions in compiled context) with an option, and finally to be able to deactivate the ghost highlighting for those who really don't want it.

Also, with respect to the background color, I suggest that a less contrasted approach should be available as an option, although I don't have a strong opinion.

![othergray](https://user-images.githubusercontent.com/3601079/142016970-1ce97a11-3e95-4ee2-ba05-d4afeafcadd7.png)

Hex color of image above: 0x323232. Hex of your color: 0x525252.

**Comment by camrein** [comment]
> Also, with respect to the background color, I suggest that a less contrasted approach should be available as an option, although I don't have a strong opinion.

I agree. Thanks for sharing the color code. I've also reduced the contrast for the light theme.
![image](https://user-images.githubusercontent.com/18049310/142215605-fa58b15a-9af5-45b4-9623-1955a55cd537.png)
